# SynVisio deployment on a server with custom datasets
SynVisio is an interactive multiscale visualization tool that lets you explore the results of McScanX a popular synteny and collinearity detection toolkit.

SynVisio is a single page react web application and so can be hosted as a simple static webpage. To present your own datasets using SynVisio you can create a custom build of SynVisio and then host it on any standard webserver such as Nginx or Apache. 

###  1- Initial Setup

To get started make sure you have NodeJS installed. Then clone the main synvisio github repository and checkout the **self-hosted-module** branch.

`git clone https://github.com/kiranbandi/synvisio.git`
`cd synvisio`
`git checkout self-hosted-module`

Then install the required npm packages - 

`npm run install`

After this you can start the local server to test the system locally first - 

`npm run start`

Then navigate to [http://localhost:8080](http://localhost:8080) to see the synteny dataset of **Brassica napus (bn)** load up. 

### 2- Setup files for custom datasets

SynVisio loads datasets directly using unique identifiers in its URL. So to get your own datasets in SynVisio instead of **Brassica napus**, copy your input files to the folder **build/assets/files** inside the main synvisio folder and then name them in the following manner. 

Each dataset needs to have a unique identifier, for example lets say the identifer for your dataset is **"sample"**. Then you need to rename the McScanX collinearity output file as **sample_collinear.collinearity** and the corresponding gff file as **sample_coordinate.gff**. 

Optionally if you have a gene count track that you want to present alongside the synteny visualization you can also add your gene count file to the same folder and name it as **sample_gene_count.txt**.

If you would like to have more that one dataset avaiable you just need to follow the above naming procedure and repeat it with different unique indentifiers. 

Once your files have been placed in the correct folder, you can then view them by navigating to [http://localhost:8080/#/sample](http://localhost:8080/#/sample). Make sure to replace "sample" with the unique identifier name you used for your files. 

### 3- Setup default configuration

SynVisio by defaults loads up the unique identifier **\*bn*** which corresponds to Brassica napus. But this can be changed to the unique identifier that you used for your files. To replace this simply replace 'bn' with your unique idenitifer name in the file **defaultDataSource.js** which can be found in the folder **src/utils**.

Then you can also setup the default source and target chromosomes that are pre selected in the filter panel, for this you will need to edit the **defaultDatasetMap.js** file. It can be found in the same folder **src/utils**. You need to add an entry in the json object for your unique identifier and then simply list the chromosomes inside as an array. Use the existing the entries in the file present for **bn** as reference. 

### 4- Create a build

Once you are done adding all your datasets and setting up the default configurations check to see that the web application is working as expected by navigating to [http://localhost:8080/#/your-unique-identifier-here](http://localhost:8080/#/your-unique-identifier-here). The datasets should load up and if the source and target chromsomes are preset they should be pre selected in the filter panel. 

If everything is working as expected, you are now ready to create a static build. For this run the following command - 

`npm run build`

This will build and package the entire react application into a single bundle. Now all the assets needed for your application will be present inside the **\*build*** folder. You can move the contents of this folder into your Apache or Nginx hosting folder (which is usually at /var/www/html/ if you are on a unix based server). 

### 5- Publication Citation
Please consider citing our work if you are using images generated by SynVisio or are presenting your datasets via SynVisio in a research publication.

>Venkat Bandi, and Carl Gutwin. 2020. Interactive Exploration of Genomic Conservation. In Proceedings of the 46th Graphics Interface Conference on Proceedings of Graphics Interface 2020 (GIâ€™20). Canadian Human-Computer Communications Society, Waterloo, CAN.
